lib/base/unique_value.h

```diff
@@ -0,0 +1,112 @@
+#ifndef LIB_BASE_UNIQUE_VALUE_H_
+#define LIB_BASE_UNIQUE_VALUE_H_
+
+#include <functional>
+#include <optional>
+
+#include "lib/base/invariant.h"
+
+namespace base {
+
+// Type providing std::unique_ptr<>-like functionality for immutable values.
+// The contents of UniqueValue<> may be replaced, but not modified in place.
+//
+// Unlike pointers where nullptr is a special value that indicates invalid
+// contents, every possible value a value object can hold may be valid. This
+// requires additional storage to track validity of the contained value. Despite
+// the required extra state, this is still more efficient than storing values in
+// heap allocated space and behaves more gracefully under memory pressure.
+//
+// In addition to the type of the wrapped value, users must provide a cleanup
+// function object to be called whenever a UniqueValue containing a valid value
+// is destroyed.
+//
+// CleanupFunctionObjectT is a function object implementing the call operator.
+// In C++20 lambdas can be declared in unevaluated contexts 'decltype([](const T
+// value){})' sufficies to declare a cleanup function object. Prior to C++20 a
+// type must be defined:
+//
+// struct CleanupFunctionObject {
+//     void operator()(const T value){
+//         // cleanup 'value', only called with valid 'value'.
+//     };
+// };
+//
+// Example usage:
+//
+// struct CloseFunctionObject { void operator()(const int fd) { close(fd); } };
+// using UniqueFileDescriptor = UniqueValue<int, CloseFunctionObject>;
+// {
+//     UniqueFileDescriptor fd(open("/some/path", FLAGS));
+//     DoSomethingWithFd(GetValue(fd));  // Free function style value access.
+//     DoSomethingElseWithFd(*fd);  // std::unique_ptr style value access.
+// }  // 'fd' leaves scope and is automatically closed!
+//
+// See /lib/posix/unique_file_descriptor.h for a more complete example.
+//
+template <typename T, typename CleanupFunctionObjectT>
+class UniqueValue {
+ public:
+  // Like std::unique_ptr<> default construction results in an empty object.
+  constexpr UniqueValue() = default;
```

Comment:
I get that we're modeling unique_ptr here, but the primary use case is a little different here. Do we ever want to allow empty values to be constructed? For things like file descriptors that we want to ensure get cleaned up, do we lose anything by insisting that the container always holds a valid value? Wouldn't this be more efficient too?

Reply:
You've actually discovered a deficiency of C++. Types without default constructors can cause all sorts of problems because there are many containers that need to reserve space before values are available. For example, std::vector<> cannot be resized without a default constructor. Non-trivial constructors also run in to trouble in classes that have members without default constructors.

The C++ solution is related to std::move. Movable types are responsible for tracking enough information to inhibit non-trivial destruction work when their contents are moved to another storage location. Typically objects then default construct to a state similar or identical to the "moved from" state.

This probably seems a bit backwards to you coming from a nice language like rust.